File: packages/sdk/hush-py/src/clawdstrike/guards/prompt_injection.py
"""Prompt injection guard - detects common prompt-injection patterns in untrusted text.

This is a pure-Python implementation of the heuristic + statistical detection layers,
matching the Rust PromptInjectionGuard's behavior.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from enum import IntEnum
from typing import Any

from clawdstrike.guards.base import (
    Action,
    Guard,
    GuardContext,
    GuardResult,
    Severity,
)


class PromptInjectionLevel(IntEnum):
    """Detection confidence level."""

    SAFE = 0
    SUSPICIOUS = 1
    HIGH = 2
    CRITICAL = 3


# Heuristic signal patterns (case-insensitive)
_SIGNALS: list[tuple[str, str, int]] = [  # noqa: E501
    # (signal_id, pattern, score_contribution)
    ("ignore_previous_instructions", r"(?i)ignore\s+(all\s+)?(previous|prior|above|earlier)\s+(instructions|prompts|rules)", 40),  # noqa: E501
    ("reveal_system_prompt", r"(?i)(reveal|show|display|output|print|repeat)\s+(the\s+)?(system\s+)?(prompt|instructions|rules)", 35),  # noqa: E501
    ("new_instructions", r"(?i)(new|updated|override|replacement)\s+(system\s+)?(instructions|prompt|rules)", 30),  # noqa: E501
    ("you_are_now", r"(?i)you\s+are\s+now\s+(a|an|the|my)\s+(unrestricted|unfiltered|uncensored|jailbroken?|evil|new|different)\s", 25),  # noqa: E501
    ("do_anything_now", r"(?i)(DAN|do\s+anything\s+now)", 30),
    ("jailbreak_prefix", r"(?i)(ignore\s+safety|bypass\s+restrictions|disable\s+filter)", 35),
    ("role_play_override", r"(?i)pretend\s+(you\s+are|to\s+be|you're)\s+", 20),
    ("encoding_evasion", r"(?i)(base64|rot13|hex|unicode)\s*(encode|decode|convert)", 15),
    ("delimiter_injection", r"(?i)(<\|im_start\|>|<\|im_end\|>|\[INST\]|\[/INST\]|<\|system\|>|<\|user\|>|<\|assistant\|>)", 25),  # noqa: E501
    ("exfiltration_request", r"(?i)(send|transmit|exfiltrate|post)\s+(to|data|secrets|keys)\s+(http|url|endpoint|server)", 40),  # noqa: E501
]

_COMPILED_SIGNALS = [(sid, re.compile(pat), score) for sid, pat, score in _SIGNALS]


def _detect(
    text: str, max_scan_bytes: int = 200_000,
) -> tuple[PromptInjectionLevel, int, list[str]]:
    """Run heuristic detection on text prefix.

    Returns (level, score, list_of_signal_ids).
    """
    scanned = text[:max_scan_bytes]
    total_score = 0
    matched_signals: list[str] = []

    for signal_id, compiled, score in _COMPILED_SIGNALS:
        if compiled.search(scanned):
            total_score += score
            matched_signals.append(signal_id)

    if total_score >= 60:
        level = PromptInjectionLevel.CRITICAL
    elif total_score >= 35:
        level = PromptInjectionLevel.HIGH
    elif total_score >= 15:
        level = PromptInjectionLevel.SUSPICIOUS
    else:
        level = PromptInjectionLevel.SAFE

    return level, total_score, matched_signals


@dataclass
class PromptInjectionConfig:
    """Configuration for PromptInjectionGuard."""

    enabled: bool = True
    warn_at_or_above: str = "suspicious"
    block_at_or_above: str = "high"
    max_scan_bytes: int = 200_000


def _level_from_str(s: str) -> PromptInjectionLevel:
    mapping = {
        "safe": PromptInjectionLevel.SAFE,
        "suspicious": PromptInjectionLevel.SUSPICIOUS,
        "high": PromptInjectionLevel.HIGH,
        "critical": PromptInjectionLevel.CRITICAL,
    }
    return mapping.get(s.lower(), PromptInjectionLevel.HIGH)


class PromptInjectionGuard(Guard):
    """Guard that evaluates prompt-injection risk for untrusted text.

    Handles custom actions with type 'untrusted_text' or 'hushclaw.untrusted_text'.
    """

    UNTRUSTED_TEXT_KINDS = {"untrusted_text", "hushclaw.untrusted_text"}

    def __init__(self, config: PromptInjectionConfig | None = None) -> None:
        self._config = config or PromptInjectionConfig()
        self._warn_level = _level_from_str(self._config.warn_at_or_above)
        self._block_level = _level_from_str(self._config.block_at_or_above)

    @property
    def name(self) -> str:
        return "prompt_injection"

    def handles(self, action: Action) -> bool:
        if not self._config.enabled:
            return False
        custom_type: str | None = getattr(action, "custom_type", None)
        return (
            action.action_type == "custom"
            and custom_type is not None
            and custom_type in self.UNTRUSTED_TEXT_KINDS
        )

    def _extract_text(self, action: Action) -> str | None:
        """Extract text from action payload."""
        data: dict | None = getattr(action, "custom_data", None)
        if data is None:
            return None
        text = data.get("text")
        if isinstance(text, str):
            return text
        return None

    def check(self, action: Action, context: GuardContext) -> GuardResult:
        if not self._config.enabled:
            return GuardResult.allow(self.name)

        if not self.handles(action):
            return GuardResult.allow(self.name)

        text = self._extract_text(action)
        if text is None:
            return GuardResult.block(
                self.name,
                Severity.ERROR,
                "Invalid untrusted_text payload: missing text field",
            )

        level, score, signals = _detect(text, self._config.max_scan_bytes)

        details: dict[str, Any] = {
            "level": level.name.lower(),
            "score": score,
            "signals": signals,
        }
        custom_data: dict | None = getattr(action, "custom_data", None)
        source = custom_data.get("source") if custom_data else None
        if source:
            details["source"] = source

        if level >= self._block_level:
            sev = {
                PromptInjectionLevel.CRITICAL: Severity.CRITICAL,
                PromptInjectionLevel.HIGH: Severity.ERROR,
            }.get(level, Severity.ERROR)
            return GuardResult.block(
                self.name,
                sev,
                f"Untrusted text contains prompt-injection signals ({level.name})",
            ).with_details(details)

        if level >= self._warn_level:
            return GuardResult.warn(
                self.name,
                f"Untrusted text contains prompt-injection signals ({level.name})",
            ).with_details(details)

        return GuardResult.allow(self.name)


__all__ = [
    "PromptInjectionGuard",
    "PromptInjectionConfig",
    "PromptInjectionLevel",
]

File: packages/sdk/hush-go/guards/prompt_injection.go
package guards

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"strings"

	"github.com/backbay/clawdstrike-go/policy"
)

// Default prompt injection thresholds.
const (
	DefaultWarnThreshold  = 0.5
	DefaultBlockThreshold = 0.8
	DefaultMaxScanBytes   = 200 * 1024 // 200KB
)

// InjectionSeverity classifies the severity of a prompt injection attempt.
type InjectionSeverity int

const (
	InjectionSafe       InjectionSeverity = iota
	InjectionSuspicious
	InjectionHigh
	InjectionCritical
)

func (s InjectionSeverity) String() string {
	switch s {
	case InjectionSafe:
		return "safe"
	case InjectionSuspicious:
		return "suspicious"
	case InjectionHigh:
		return "high"
	case InjectionCritical:
		return "critical"
	default:
		return "unknown"
	}
}

// Suspicious keywords and patterns for heuristic detection.
var suspiciousPatterns = []string{
	"ignore previous",
	"ignore all previous",
	"ignore the above",
	"disregard previous",
	"disregard all previous",
	"disregard the above",
	"forget previous",
	"forget all previous",
	"new instructions",
	"new instruction",
	"override instructions",
	"system prompt",
	"you are now",
	"act as",
	"pretend to be",
	"jailbreak",
	"ignore safety",
	"ignore guidelines",
	"bypass restrictions",
	"bypass safety",
	"do not refuse",
	"do anything",
	"no restrictions",
	"developer mode",
	"dan mode",
}

// PromptInjectionGuard detects prompt injection in untrusted text.
type PromptInjectionGuard struct {
	warnThreshold  float64 // 0 means "use default" per Go zero-value convention
	blockThreshold float64 // 0 means "use default" per Go zero-value convention
	maxScanBytes   int     // 0 means "use default" per Go zero-value convention
}

// NewPromptInjectionGuard creates a guard with the given config. Nil config uses defaults.
func NewPromptInjectionGuard(cfg *policy.PromptInjectionConfig) *PromptInjectionGuard {
	g := &PromptInjectionGuard{
		warnThreshold:  DefaultWarnThreshold,
		blockThreshold: DefaultBlockThreshold,
		maxScanBytes:   DefaultMaxScanBytes,
	}
	if cfg != nil {
		if cfg.WarnThreshold > 0 {
			g.warnThreshold = cfg.WarnThreshold
		}
		if cfg.BlockThreshold > 0 {
			g.blockThreshold = cfg.BlockThreshold
		}
		if cfg.MaxScanBytes > 0 {
			g.maxScanBytes = cfg.MaxScanBytes
		}
	}
	return g
}

func (g *PromptInjectionGuard) Name() string { return "prompt_injection" }

func (g *PromptInjectionGuard) Handles(action GuardAction) bool {
	return action.Type == "custom" && action.CustomType == "untrusted_text"
}

func (g *PromptInjectionGuard) Check(action GuardAction, ctx *GuardContext) GuardResult {
	text, ok := action.CustomData.(string)
	if !ok {
		// Not a string — can't scan. Fail-closed: block.
		return Block(g.Name(), Error, "untrusted_text data is not a string")
	}

	// Truncate to max scan bytes
	if len(text) > g.maxScanBytes {
		text = text[:g.maxScanBytes]
	}

	// Compute fingerprint
	hash := sha256.Sum256([]byte(text))
	fingerprint := hex.EncodeToString(hash[:])

	// Heuristic scoring
	score := heuristicScore(text)
	severity := scoreToSeverity(score)

	details := map[string]interface{}{
		"score":       score,
		"severity":    severity.String(),
		"fingerprint": fingerprint,
	}

	if score >= g.blockThreshold {
		return Block(g.Name(), Critical,
			fmt.Sprintf("prompt injection detected (score=%.2f)", score)).
			WithDetails(details)
	}

	if score >= g.warnThreshold {
		return Warn(g.Name(),
			fmt.Sprintf("suspicious text detected (score=%.2f)", score)).
			WithDetails(details)
	}

	return Allow(g.Name()).WithDetails(details)
}

// heuristicScore calculates a 0.0–1.0 score based on suspicious pattern matches.
func heuristicScore(text string) float64 {
	lower := strings.ToLower(text)
	matches := 0
	for _, pat := range suspiciousPatterns {
		if strings.Contains(lower, pat) {
			matches++
		}
	}
	if matches == 0 {
		return 0.0
	}
	// Normalize: 3+ matches → 1.0
	score := float64(matches) / 3.0
	if score > 1.0 {
		score = 1.0
	}
	return score
}

func scoreToSeverity(score float64) InjectionSeverity {
	switch {
	case score >= 0.8:
		return InjectionCritical
	case score >= 0.5:
		return InjectionHigh
	case score > 0.0:
		return InjectionSuspicious
	default:
		return InjectionSafe
	}
}
